lstchain/reco/utils.py

```diff
@@ -778,6 +781,127 @@ def get_geomagnetic_delta(zen, az, geomag_dec=None, geomag_inc=None, time=None):
+    
+    efftime, _ = get_effective_time(data)
+
+    nbins = 1 + int( (np.log10(max_intensity) - np.log10(min_intensity)) / step)
+    bins = np.logspace(np.log10(min_intensity), 
+                       np.log10(max_intensity), nbins)
+    binwidth = np.diff(bins)
+    nevents, _ = np.histogram(data['intensity'], bins=bins)
+    nevents = nevents.astype('float') / (binwidth*efftime.to_value(u.s))
+    bincenters = (bins[1:]*bins[:-1])**0.5 # geometrical mean (log bin center)
+ 
+    peaks, properties = find_peaks(np.log10(nevents), 
+                                   prominence=0.04, # ~10% in log10 scale
+                                   width=int(0.1/step)) # ~25% in log10 scale
+
+    # If no peak is found, nans are returned (except for the histogram data):
+    if len(peaks) == 0:
+        return np.nan, np.nan, np.nan, np.nan, bincenters, nevents
+    
+    xmax = bincenters[peaks.max()] # The peak at highest intensity (spurious peaks sometimes at low values)
+    ymax = nevents[peaks.max()]
+
+    # Find the intensity for which 50% of peak dR/dI is reached, moving down from the peak, in finer steps
+    # for better precision:
+    finestep = 0.001 # in log10(intensity)
+    xx = np.logspace(np.log10(min_intensity), 
+                     np.log10(xmax), 
+                     1+int((np.log10(xmax) - np.log10(min_intensity))/finestep))[::-1]
+
+    # Linear interpolation in dR/dI (differential rate) vs. log10(intensity/p.e.):    
+    yy = np.interp(np.log10(xx), np.log10(bincenters), nevents)
+
+    x50 = np.nan
+    y50 = np.nan
+    for xxx, yyy in zip(xx, yy):
+        if yyy < 0.5 * ymax:
+            x50 = xxx
+            y50 = np.interp(np.log10(xxx), np.log10(bincenters), nevents)
+            break
+
+    # If the 50% of peak rate is not found, x50 and y50 will be returned as nan
+    # In some cases this may be because the cosmic-ray dR/dI peak merges with a 
+    # lower-intensity peak from star-illuminated pixels, meteors, satellites...
+    # The data may still be usable, and in that case we should probably apply the
+    # default low intensity cut (50 p.e.). If data turn out to have some more 
+    # serious problem, that should be determined elsewhere.
+    
+    return xmax, ymax, x50, y50, bincenters, nevents
+
+
+def get_intensity_cut(data):
+    """
+    Obtain from a dl2 data table the recommended minimum intensity
+    cut to be applied in the event selection as a "software trigger"
+    to improve the agreement between data and MC.
+    
+    Parameters
+    ----------
+    data: pandas DataFrame or astropy.table.QTable of dl2 events.
+
+    Returns
+    -------
+    intensity_cut: (p.e.) Recommended intensity cut (both in data and MC)
+    to achieve a good match between data and MC
+    """
+
+    # Factor by which we multiply the "intensity threshold" (intensity at which
+    # 50% of peak rate dR/dI is reached) to obtain the (minimum) intensity cut. 
+    factor = 1.3 
```

Comment:
maybe we could add a plot to this review to justify this factor in case it needs to be revisited in the future